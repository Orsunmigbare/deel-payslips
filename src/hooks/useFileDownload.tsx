import { useState } from "react";
import { Capacitor } from "@capacitor/core"
import { Filesystem, Directory } from '@capacitor/filesystem';
import { FileOpener } from "@capacitor-community/file-opener"
import { base64FromPath, downloadFile } from "src/utilities/fileDownload";


export const useFileDownload = () => {
    const [loading, setLoading] = useState(false);

    const download = async (url: string, filename: string, onComplete?: () => void, onError?: () => void) => {
        if (Capacitor.isNativePlatform()) {
            try {
                setLoading(true);
                const base64 = await base64FromPath(url);
                const saveFile = await Filesystem.writeFile({
                    path: filename,
                    data: base64,
                    directory: Directory.Data
                })

                await FileOpener.open({
                    filePath: saveFile.uri
                })
                setLoading(false);
                onComplete?.();
            } catch (err) {
                setLoading(false);
                onError?.()
            }

        } else {
            downloadFile(url, filename);
        }
    }

    return {
        download,
        loading,
    }
}