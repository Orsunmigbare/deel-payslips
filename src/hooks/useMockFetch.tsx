import { useState, useEffect } from "react";
import { fetchPayslip, fetchPayslips } from "utilities/mockAPI"


type MockFetchState = {
    loading: boolean;
    data?: any
}

export const useMockFetch = (type: string, id?: string) => {
    const [state, setState] = useState<MockFetchState>({
        loading: false,
    })

    const fetchData = async (type: string) => {
        setState({ ...state, loading: true })
        let response;
        if (type === "payslip" && id) {
            response = await fetchPayslip(id);
        } else {
            response = await fetchPayslips()
        }
        setState({ loading: false, data: response.data })
    }

    useEffect(() => {
        fetchData(type)
    }, [])

    return {
        loading: state.loading,
        data: state.data
    }
}