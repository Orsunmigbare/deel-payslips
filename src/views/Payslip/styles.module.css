.payslip {
    display: flex;
    flex-direction: column;
}

.subHeading {
    font-weight: bold;
    font-size: 16px;
    margin: 10px 0;
}

.preview {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 30px 0;
    padding: 30px;
    height: 60vh;
    background-color: white;
    border: 1px solid #e8ebf2;
}

.previewTitle {
    font-size: 14px;
    width: 100%;
    color: #444bc5;
    font-weight: bold;
}

.preview .imgCont {
    overflow: hidden;
}


.footer {
    padding: 20px 0;
    width: 100%;
}

.button {
    width: 100%;
}