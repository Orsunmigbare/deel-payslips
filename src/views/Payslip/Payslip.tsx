import {
    IonPage,
    IonContent,
    IonHeader,
    IonTitle,
    IonToolbar,
    IonFooter,
    IonButtons,
    IonBackButton,
    IonImg,
    IonButton,
    IonSpinner,
    useIonToast,
} from '@ionic/react';
import { useParams } from 'react-router';
import { Payslip as IPayslip } from 'src/types/api';
import { useMockFetch } from 'src/hooks/useMockFetch';
import { useFileDownload } from 'src/hooks/useFileDownload';
import { Conditional } from 'src/components/Conditional/Conditional';
import { showToast } from 'src/utilities/toast';
import styles from "./styles.module.css"



type Param = {
    id: string
}

type FetchResponse = {
    loading: boolean;
    data: IPayslip
}




function Payslip() {
    let { id } = useParams<Param>();
    const [present] = useIonToast();
    const { loading, data: slip }: FetchResponse = useMockFetch("payslip", id);
    const { download, loading: downloading } = useFileDownload()

    const onDownload = (slip: IPayslip) => {
        const filename = `${new Date(slip.fromDate).toDateString()}-${new Date(slip.toDate).toDateString()}.png`
        download(
            slip.file,
            filename,
            () => showToast("Payslip Saved Successfully!", present),
            () => showToast("An error occurred saving payslips", present)
        )
    }

    return (
        <IonPage>
            <IonHeader>
                <IonToolbar>
                    <IonButtons slot="start">
                        <IonBackButton defaultHref="/" ></IonBackButton>
                    </IonButtons>
                    <IonTitle>Payslip</IonTitle>
                </IonToolbar>
            </IonHeader>
            <IonContent className="ion-padding">
                <Conditional data={slip} loading={loading}>
                    <div className={styles.payslip}>
                        <h2 className={styles.subHeading}> Payslip ID : {slip?.id}</h2>
                        <div className={styles.preview}>
                            <h3 className={styles.previewTitle}> Duration : {new Date(slip?.fromDate).toDateString()} - {new Date(slip?.toDate).toDateString()} </h3>
                            <div className={styles.imgCont}>
                                <IonImg
                                    src={slip?.file} alt="payslip"
                                ></IonImg>
                            </div>
                        </div>
                    </div>
                </Conditional>

            </IonContent>
            <Conditional loading={false} data={slip} >
                <IonFooter className="ion-padding">
                    <div className={styles.footer}>
                        <IonButton
                            disabled={downloading}
                            onClick={() => onDownload(slip)}
                            className={styles.button}>
                            {downloading ? <IonSpinner /> : "Download Payslip"}
                        </IonButton>
                    </div>
                </IonFooter>
            </Conditional >
        </IonPage >
    )
}

export default Payslip;
