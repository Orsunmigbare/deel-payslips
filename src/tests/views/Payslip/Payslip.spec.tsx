import "@testing-library/jest-dom"
import { render, fireEvent } from '@testing-library/react';
import { useFileDownload } from "hooks/useFileDownload";
import { useMockFetch } from "hooks/useMockFetch";
import Payslip from "views/Payslip/Payslip";
import { sampleSlip } from "./data";
import { useParams } from 'react-router';


jest.mock("react-router", () => (
    {
        ...(jest.requireActual("react-router")),
        useParams: jest.fn()
    }
))
jest.mock("../../../hooks/useFileDownload", () => (
    {
        ...(jest.requireActual("../../../hooks/useFileDownload")),
        useFileDownload: jest.fn()
    }
))
jest.mock("../../../hooks/useMockFetch", () => (
    {
        ...(jest.requireActual("../../../hooks/useMockFetch")),
        useMockFetch: jest.fn()
    }
));


afterEach(() => {
    jest.resetAllMocks();
});
describe("Payslip", () => {
    it("renders correctly", () => {
        (useMockFetch as jest.Mock).mockReturnValue({ loading: false, data: sampleSlip });
        (useParams as jest.Mock).mockReturnValue({ id: sampleSlip.id });
        (useFileDownload as jest.Mock).mockReturnValue({
            download: jest.fn(),
            loading: false
        });
        const idTitle = `Payslip ID : ${sampleSlip.id}`,
            durationTitle = `Duration : ${new Date(sampleSlip.fromDate).toDateString()} - ${new Date(sampleSlip.toDate).toDateString()}`

        const screen = render(<Payslip />);
        expect(screen.getByText(idTitle)).toBeInTheDocument();
        expect(screen.getByText(durationTitle)).toBeInTheDocument();
        expect(screen.getByAltText("payslip")).toHaveAttribute("src", sampleSlip.file);
        expect(screen).toMatchSnapshot()
    });

    it("calls useFileDownload.download when the download button is clicked", () => {
        (useMockFetch as jest.Mock).mockReturnValue({ loading: false, data: sampleSlip });
        (useParams as jest.Mock).mockReturnValue({ id: sampleSlip.id });
        (useFileDownload as jest.Mock).mockReturnValue({
            download: jest.fn(),
            loading: false
        });
        const filename = `${new Date(sampleSlip.fromDate).toDateString()}-${new Date(sampleSlip.toDate).toDateString()}.png`
        const { getByText } = render(<Payslip />);
        fireEvent.click(getByText("Download Payslip"));
        expect(useFileDownload().download).toHaveBeenCalledTimes(1);
        expect(useFileDownload().download).toHaveBeenCalledWith(sampleSlip.file, filename, expect.any(Function), expect.any(Function));
    })
})