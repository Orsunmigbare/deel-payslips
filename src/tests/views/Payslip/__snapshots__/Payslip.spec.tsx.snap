// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payslip renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ion-page"
      >
        <ion-header>
          <!---->
          <ion-toolbar>
            <ion-buttons
              slot="start"
            >
              <ion-back-button
                default-href="/"
              />
            </ion-buttons>
            <ion-title>
              Payslip
            </ion-title>
          </ion-toolbar>
        </ion-header>
        <ion-content
          class="ion-padding"
        >
          <div
            class="payslip"
          >
            <h2
              class="subHeading"
            >
               Payslip ID : 
              61f4a809a6c0b5d0e1840c11
            </h2>
            <div
              class="preview"
            >
              <h3
                class="previewTitle"
              >
                 Duration : 
                Tue Feb 01 2022
                 - 
                Tue Mar 01 2022
                 
              </h3>
              <div
                class="imgCont"
              >
                <ion-img
                  alt="payslip"
                  src="https://www.personalfinancelab.com/wp-content/uploads/h5p/content/13/images/image-5e502af138e62.PNG"
                />
              </div>
            </div>
          </div>
        </ion-content>
        <ion-footer
          class="ion-padding"
        >
          <!---->
          <div
            class="footer"
          >
            <ion-button
              class="button"
              disabled="false"
            >
              Download Payslip
            </ion-button>
          </div>
        </ion-footer>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ion-page"
    >
      <ion-header>
        <!---->
        <ion-toolbar>
          <ion-buttons
            slot="start"
          >
            <ion-back-button
              default-href="/"
            />
          </ion-buttons>
          <ion-title>
            Payslip
          </ion-title>
        </ion-toolbar>
      </ion-header>
      <ion-content
        class="ion-padding"
      >
        <div
          class="payslip"
        >
          <h2
            class="subHeading"
          >
             Payslip ID : 
            61f4a809a6c0b5d0e1840c11
          </h2>
          <div
            class="preview"
          >
            <h3
              class="previewTitle"
            >
               Duration : 
              Tue Feb 01 2022
               - 
              Tue Mar 01 2022
               
            </h3>
            <div
              class="imgCont"
            >
              <ion-img
                alt="payslip"
                src="https://www.personalfinancelab.com/wp-content/uploads/h5p/content/13/images/image-5e502af138e62.PNG"
              />
            </div>
          </div>
        </div>
      </ion-content>
      <ion-footer
        class="ion-padding"
      >
        <!---->
        <div
          class="footer"
        >
          <ion-button
            class="button"
            disabled="false"
          >
            Download Payslip
          </ion-button>
        </div>
      </ion-footer>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
